packages/server/src/router.ts

```diff
@@ -150,11 +150,30 @@ const defaultTransformer: CombinedDataTransformer = {
   input: { serialize: (obj) => obj, deserialize: (obj) => obj },
   output: { serialize: (obj) => obj, deserialize: (obj) => obj },
 };
-export type MiddlewareFunction<TContext> = (opts: {
+
+export const middlewareMarker = Symbol('middlewareMarker');
+export interface MiddlewareResult {
+  /**
+   * All middlewares should pass through their `next()`'s output.
+   * Requiring this marker makes sure that can't be forgotten at compile-time.
+   */
+  readonly marker: typeof middlewareMarker;
+  output: {};
+}
+export interface MiddlewareFunctionOpts<TContext> {
   ctx: TContext;
   type: ProcedureType;
   path: string;
-}) => Promise<void> | void;
+  input: unknown;
+}
+export type MiddlewareNextFn = () => Promise<MiddlewareResult>;
+
+export type MiddlewareFunction<TContext> = (
```

Comment:
Is there any reason to actually export these? 🤔

Reply:
No, you're right, these can be reduced down. The only type we really need to import elsewhere is `MiddlewareFunction`, I'll un-export the rest.